### ISSUE TITLE
Unexpected error for class using a combination of "keyof", generics and unions

### ISSUE THREAD

**Issue by ashtonsix** [microsoft/TypeScript#47629]
# Bug Report

### 🔎 Search Terms

`keyof`,  `generic`, `union`

### 🕗 Version & Regression Information

Occurs in version 4.5.4, and this is the only version I have tried.

### ⏯ Playground Link

https://www.typescriptlang.org/play?#code/C4TwDgpgBAqlC8UDeBDAXFADAXygH2XSgEZsAoMgYwBsUBnOqAIQHsAPAHgDUoI3gIAOwAmjGAD5kZKFABuAfgwBrCCBYAzKF2lRKLQXWAAnAK6VgLIwApZGLgEpk5cmWoRgUAEYZWnCQihBCAB3ZnYrVAxSezIgA

### 💻 Code

```ts
type U = {a: 0} | {a: 1}

class Box<V extends U> {
  v?: keyof V
  constructor(v: V) {}
}

let b: Box<U> = new Box({a: 1})
```

### 🙁 Actual behavior

When trying to assign `new Box({a: 1})` to `Box<U>` this error occurs:

```
Type 'Box<{ a: 1; }>' is not assignable to type 'Box<U>'.
  Type 'U' is not assignable to type '{ a: 1; }'.
    Type '{ a: 0; }' is not assignable to type '{ a: 1; }'.
      Types of property 'a' are incompatible.
        Type '0' is not assignable to type '1'.(2322)
```

This error does not occur if the line `v?: keyof V` is removed.

### 🙂 Expected behavior

The code should not produce an error, and `typeof b.v` should equal `"a"`.

**Comment by RyanCavanaugh** [comment]
This is a correct error.
```ts
type U = {a: 0} | {a: 1}

class Box<V extends U> {
  v?: keyof V
  constructor(v: V) {}
  do_set(m: keyof V) {
    this.v = m;
  }
  do_get() {
    return this.v!;
  }
}

const j = new Box({a: 1, b: 2});
j.do_set("b");

// Alleged: this is OK
let b: Box<U> = j;
// Unsound: value 'b' inhabits 'a'
const a: "a" = b.do_get();
```

**Comment by ashtonsix** [comment]
Gotcha, thank you for correcting my misunderstanding.
